pkg/gardenlet/bootstrap/certificate/certificate_rotation.go

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certificate
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/gardener/gardener/pkg/gardenlet/bootstrap/util"
+	"github.com/gardener/gardener/pkg/logger"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/retry"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// certificateWaitTimeout controls the amount of time we wait for the certificate
+// approval in one iteration.
+var certificateWaitTimeout = 15 * time.Minute
+
+// Manager can be used to schedule the certificate rotation for the Gardenlet's Garden cluster client certificate
+type Manager struct {
+	stopCh chan struct{}
+
+	seedName               string
+	clientMap              clientmap.ClientMap
+	gardenClientConnection config.GardenClientConnection
+
+	logger logrus.FieldLogger
+	// Set to time.Now but can be stubbed out for testing
+	now func() time.Time
+}
+
+// NewCertificateRotation creates a certificate manager that can be used to rotate gardenlet's client certificate for the Garden cluster
+func NewCertificateManager(gardenClientConnection config.GardenClientConnection, seedConfig *config.SeedConfig, clientMap clientmap.ClientMap) *Manager {
```

Comment:
Can you pass a pointer here?
```suggestion
func NewCertificateManager(gardenClientConnection *config.GardenClientConnection, seedConfig *config.SeedConfig, clientMap clientmap.ClientMap) *Manager {
```